File: modules/common/src/main/java/org/opencastproject/rest/BulkOperationResult.java
/*
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 *
 * The Apereo Foundation licenses this file to you under the Educational
 * Community License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at:
 *
 *   http://opensource.org/licenses/ecl2.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package org.opencastproject.rest;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;

import java.util.ArrayList;
import java.util.List;

/**
 * This class is used to store the results of a bulk operation on an endpoint and to easily return those results.
 */
public class BulkOperationResult {
  public static final String OK_KEY = "ok";
  public static final String ACCEPTED_KEY = "accepted";
  public static final String BAD_REQUEST_KEY = "badRequest";
  public static final String UNAUTHORIZED_KEY = "unauthorized";
  public static final String NOT_FOUND_KEY = "notFound";
  public static final String ERROR_KEY = "error";

  private final List<String> ok = new ArrayList<>();
  private final List<String> accepted = new ArrayList<>();
  private final List<String> badRequest = new ArrayList<>();
  private final List<String> unauthorized = new ArrayList<>();
  private final List<String> notFound = new ArrayList<>();
  private final List<String> serverError = new ArrayList<>();

  public void addOk(String id) {
    ok.add(id);
  }

  public void addAccepted(String id) {
    accepted.add(id);
  }

  public void addBadRequest(String id) {
    badRequest.add(id);
  }

  public void addNotFound(String id) {
    notFound.add(id);
  }

  public void addServerError(String id) {
    serverError.add(id);
  }

  public void addOk(Long id) {
    addOk(Long.toString(id));
  }

  public void addBadRequest(Long id) {
    addBadRequest(Long.toString(id));
  }

  public void addUnauthorized(String id) {
    unauthorized.add(id);
  }

  public void addNotFound(Long id) {
    addNotFound(Long.toString(id));
  }

  public void addServerError(Long id) {
    addServerError(Long.toString(id));
  }

  public String toJson() {
    JsonObject bulkOperationResult = new JsonObject();
    bulkOperationResult.add(OK_KEY, toJsonArray(ok));
    bulkOperationResult.add(ACCEPTED_KEY, toJsonArray(accepted));
    bulkOperationResult.add(BAD_REQUEST_KEY, toJsonArray(badRequest));
    bulkOperationResult.add(NOT_FOUND_KEY, toJsonArray(notFound));
    bulkOperationResult.add(UNAUTHORIZED_KEY, toJsonArray(unauthorized));
    bulkOperationResult.add(ERROR_KEY, toJsonArray(serverError));
    return bulkOperationResult.toString();
  }

  private static JsonArray toJsonArray(List<String> ids) {
    JsonArray array = new JsonArray();
    ids.forEach(array::add);
    return array;
  }
}

File: modules/common/src/main/java/org/opencastproject/rest/BulkOperationResult.kt
/**
 * Licensed to The Apereo Foundation under one or more contributor license
 * agreements. See the NOTICE file distributed with this work for additional
 * information regarding copyright ownership.
 *
 *
 * The Apereo Foundation licenses this file to you under the Educational
 * Community License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License
 * at:
 *
 * http://opensource.org/licenses/ecl2.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 */

package org.opencastproject.rest

import org.apache.commons.io.IOUtils
import org.json.simple.JSONArray
import org.json.simple.JSONObject
import org.json.simple.parser.JSONParser
import org.json.simple.parser.ParseException

import java.io.IOException
import java.io.InputStream
import java.io.StringWriter

/**
 * This class is used to store the results of a bulk operation on an endpoint and to easily return those results.
 */
class BulkOperationResult {

    var oks = JSONArray()
        private set
    var badRequests = JSONArray()
        private set
    var unauthorized = JSONArray()
        private set
    var notFound = JSONArray()
        private set
    var serverError = JSONArray()
        private set

    fun addOk(id: String) {
        oks.add(id)
    }

    fun addBadRequest(id: String) {
        badRequests.add(id)
    }

    fun addNotFound(id: String) {
        notFound.add(id)
    }

    fun addServerError(id: String) {
        serverError.add(id)
    }

    fun addOk(id: Long?) {
        addOk(java.lang.Long.toString(id!!))
    }

    fun addBadRequest(id: Long?) {
        addBadRequest(java.lang.Long.toString(id!!))
    }

    fun addUnauthorized(id: String) {
        unauthorized.add(id)
    }

    fun addNotFound(id: Long?) {
        addNotFound(java.lang.Long.toString(id!!))
    }

    fun addServerError(id: Long?) {
        addServerError(java.lang.Long.toString(id!!))
    }

    fun toJson(): String {
        val bulkOperationResult = JSONObject()
        bulkOperationResult[OK_KEY] = oks
        bulkOperationResult[BAD_REQUEST_KEY] = badRequests
        bulkOperationResult[NOT_FOUND_KEY] = notFound
        bulkOperationResult[UNAUTHORIZED_KEY] = unauthorized
        bulkOperationResult[ERROR_KEY] = serverError
        return bulkOperationResult.toJSONString()
    }

    @Throws(IOException::class, ParseException::class)
    fun fromJson(jsonContent: InputStream) {
        val writer = StringWriter()
        IOUtils.copy(jsonContent, writer)
        val parser = JSONParser()
        val result = parser.parse(writer.toString()) as JSONObject
        this.oks = result[OK_KEY] as JSONArray
        this.badRequests = result[BAD_REQUEST_KEY] as JSONArray
        this.notFound = result[NOT_FOUND_KEY] as JSONArray
        this.unauthorized = result[UNAUTHORIZED_KEY] as JSONArray
        this.serverError = result[ERROR_KEY] as JSONArray
    }

    companion object {
        val OK_KEY = "ok"
        val BAD_REQUEST_KEY = "badRequest"
        val UNAUTHORIZED_KEY = "unauthorized"
        val NOT_FOUND_KEY = "notFound"
        val ERROR_KEY = "error"
    }

}
